library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
+
+  def unary_~ : RuntimeLong = // scalastyle:ignore
+    new RuntimeLong(~lo, ~hi)
+
+  def |(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo | b.lo, a.hi | b.hi)
+
+  def &(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo & b.lo, a.hi & b.hi)
+
+  def ^(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo ^ b.lo, a.hi ^ b.hi)
+
+  // Shifts
+
+  /** Shift left */
+  def <<(n0: Int): RuntimeLong = {
+    val n = n0 & 63
+    val lo = this.lo
+
+    if (n == 0) this
+    else if (n < 32) new RuntimeLong(lo << n, (lo >>> -n) | (hi << n))
+    else new RuntimeLong(0, lo << n)
   }
 
-  /**
-   * arithmetic right shift
-   */
-  def >>(n_in: Int): RuntimeLong = {
-    val n = n_in & 63;
-
-    // Sign extend x.h
-    val negative = (x.h & SIGN_BIT_VALUE) != 0
-    val xh = if (negative) x.h | ~MASK_2 else x.h
-
-    if (n < BITS) {
-      val remBits = BITS - n
-      // FIXME IMHO the first two >> should be >>>
-      masked((x.l >> n) | (x.m << remBits),
-             (x.m >> n) | (xh  << remBits),
-             xh >> n)
-    } else if (n < BITS01) {
-      val shfBits = n - BITS
-      val remBits = BITS01 - n
-      // FIXME IMHO the first >> should be >>>
-      masked((x.m >> shfBits) | (xh << remBits),
-              xh  >> shfBits,
-             if (negative) MASK_2 else 0)
-    } else {
-      masked(xh >> (n - BITS01),
-             if (negative) MASK   else 0,
-             if (negative) MASK_2 else 0)
-    }
+  @inline
+  private def inline_<<(lo: Int, hi: Int, n: Int): (Int, Int) = {
+    if (n == 0) (lo, hi)
+    else if (n < 32) (lo << n, (lo >>> -n) | (hi << n))
+    else (0, lo << n)
+  }
+
+  /** Logical shift right */
+  def >>>(n0: Int): RuntimeLong = {
+    val n = n0 & 63
+    val hi = this.hi
 
+    if (n == 0) this
+    else if (n < 32) new RuntimeLong((lo >>> n) | (hi << -n), hi >>> n)
+    else new RuntimeLong(hi >>> n, 0)
   }
 
-  def equals(y: RuntimeLong): Boolean =
-    x.l == y.l && x.m == y.m && x.h == y.h
+  @inline
+  private def inline_>>>(lo: Int, hi: Int, n: Int): (Int, Int) = {
+    if (n == 0) (lo, hi)
+    else if (n < 32) ((lo >>> n) | (hi << -n), hi >>> n)
+    else (hi >>> n, 0)
+  }
 
-  override def equals(that: Any): Boolean = that match {
-    case y: RuntimeLong => x.equals(y)
-    case _ => false
+  /** Arithmetic shift right */
+  def >>(n0: Int): RuntimeLong = {
+    val n = n0 & 63
+    val hi = this.hi
+
+    if (n == 0) this
+    else if (n < 32) new RuntimeLong((lo >>> n) | (hi << -n), hi >> n)
+    else new RuntimeLong(hi >> n, hi >> 31)
   }
 
-  def notEquals(that: RuntimeLong): Boolean = !equals(that)
+  // Arithmetic operations
+
+  def unary_- : RuntimeLong = // scalastyle:ignore
+    inlineLongUnary_-(lo, hi)
+
+  @inline
+  private def inline_abs(lo: Int, hi: Int): (Boolean, Int, Int) = {
+    val neg = hi < 0
+    var absLo = lo
+    var absHi = hi
+    if (neg) {
+      absLo = -lo
+      absHi = if (lo != 0) ~hi else -hi
+    }
+    (neg, absLo, absHi)
```

Comment:
I'm assuming there is a very good reason why you use `var`s instead of two return paths. Please mention it.